pkg/runner/test_runner.go

```diff
@@ -6,282 +6,778 @@
+	if ss.stepRunning {
+		return
 	}
+	if ss.runErr != nil {
+		return
+	}
+	ss.stepRunning = true
+	ss.readerRunning = true
+	go tr.stepRunner(ctx, ss)
+	go tr.stepReader(ctx, ss)
+}
 
-	// rootLog is propagated to all the subsystems of the pipeline
-	rootLog := logging.GetLogger("pkg/runner")
-	fields := make(map[string]interface{})
-	fields["jobid"] = jobID
-	fields["runid"] = runID
-	rootLog = logging.AddFields(rootLog, fields)
+// emitEvent emits the specified event with the specified JSON payload (if any).
+func (ss *stepState) emitEvent(name event.Name, tgt *target.Target, payload interface{}) error {
+	var payloadJSON *json.RawMessage
+	if payload != nil {
+		payloadBytes, jmErr := json.Marshal(payload)
+		if jmErr != nil {
+			return fmt.Errorf("failed to marshal event: %w", jmErr)
+		}
+		pj := json.RawMessage(payloadBytes)
+		payloadJSON = &pj
+	}
+	errEv := testevent.Data{
+		EventName: name,
+		Target:    tgt,
+		Payload:   payloadJSON,
+	}
+	return ss.ev.Emit(errEv)
+}
 
-	log := logging.AddField(rootLog, "phase", "run")
-	testPipeline := newPipeline(logging.AddField(rootLog, "entity", "test_pipeline"), test.TestStepsBundles, test, jobID, runID, tr.timeouts)
-
-	log.Infof("setting up pipeline")
-	completedTargets := make(chan *target.Target, 1)
-	inCh := testPipeline.init(ctx)
-
-	// inject targets in the step
-	terminateInjectionCtx, terminateInjection := context.WithCancel(context.Background())
-	go func(ctx context.Context, inputChannel chan<- *target.Target) {
-		defer close(inputChannel)
-		log := logging.AddField(log, "step", "injection")
-		writer := newTargetWriter(log, tr.timeouts)
-		for _, t := range targets {
-			if err := writer.writeTimeout(ctx, inputChannel, t, tr.timeouts.MessageTimeout); err != nil {
-				log.Debugf("could not inject target %+v into first routing block: %+v", t, err)
+// stepRunner runs a test pipeline's step (the Run() method).
+func (tr *testRunner) stepRunner(ctx statectx.Context, ss *stepState) {
+	ss.log.Debugf("%s: step runner active", ss)
+	defer func() {
+		if r := recover(); r != nil {
+			tr.mu.Lock()
+			ss.stepRunning = false
+			ss.runErr = &cerrors.ErrTestStepPaniced{
+				StepName:   ss.sb.TestStepLabel,
+				StackTrace: fmt.Sprintf("%s / %s", r, debug.Stack()),
 			}
+			tr.mu.Unlock()
+			tr.safeCloseOutCh(ss)
 		}
-	}(terminateInjectionCtx, inCh)
-
-	errCh := make(chan error, 1)
-	go func() {
-		log.Infof("running pipeline")
-		errCh <- testPipeline.run(completedTargets)
 	}()
-
-	defer terminateInjection()
-	// Receive targets from the completed channel controlled by the pipeline, while
-	// waiting for termination signals or fatal errors encountered while running
-	// the pipeline.
-	for {
-		select {
-		case err := <-errCh:
-			log.Debugf("test runner terminated, returning %v", err)
-			return err
-		case t := <-completedTargets:
-			log.Infof("test runner completed target: %v", t)
-		}
+	chans := test.TestStepChannels{In: ss.inCh, Out: ss.outCh, Err: ss.errCh}
```

Comment:
Out of curiosity, why we do not initialize `test.TestStepChannels` right inside `stepState` (and instead we define three separate channels)?